### DISCUSSION TITLE
Adding new device RDZ100ZB

### DISCUSSION THREAD

**Discussion by VladislavVesely** [Koenkk/zigbee2mqtt#22739] [Devices]
Hello,

i'm trying to add unsupported thermostat from Siemens. So far i got some basic stuff working by adding (please ignore unused parts or any ugly code used during debugging). 

```
const {identify, battery, ota} = require('zigbee-herdsman-converters/lib/modernExtend');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const constants = require('zigbee-herdsman-converters/lib/constants');
const reporting  = require('zigbee-herdsman-converters/lib/reporting');
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const e = exposes.presets;
const ea = exposes.access;
const payload = reporting.payload;
const repInterval = constants.repInterval;


//export const test = async (endpoint: Zh.Endpoint, overrides?: Reporting.Override) => {
//    const p = payload('ctrlSeqeOfOper', 10, repInterval.HOUR, null, overrides);
//    await endpoint.configureReporting('hvacThermostat', p);
//};

const definition = {
    zigbeeModel: ['RDZ100'],
    model: 'RDZ100',
    vendor: 'Siemens',
    description: 'Siemens RDZ100 Wireless Thermostat',
    fromZigbee: [
        fz.thermostat,
    ],
    toZigbee: [
        tz.thermostat_local_temperature,
	tz.thermostat_local_temperature_calibration,
        tz.thermostat_occupied_heating_setpoint,
	tz.thermostat_unoccupied_heating_setpoint,
        tz.thermostat_system_mode,
    ],
    extend: [identify(), battery(), ota()],
    exposes: (device, options) => {
	const expose = [
	    e.climate()
	    .withLocalTemperature()
	    .withLocalTemperatureCalibration(-2.5, 2.5, 0.5)
	    .withSetpoint('occupied_heating_setpoint', 5, 30, 0.5)
	    .withSetpoint('unoccupied_heating_setpoint', 5, 30, 0.5)
	    .withSystemMode(['off', 'heat'])
	];
	
	expose.push(e.enum('ctrlSeqeOfOper', ea.STATE_SET, ['0', '1', '2', '3', '4']).withDescription('TODO'));

	return expose;
    },
    meta: {timeout: 100000},
    configure: async (device, coordinatorEndpoint) => {
        const endpoint = device.getEndpoint(1);

	await reporting.bind(endpoint, coordinatorEndpoint, ['genPowerCfg', 'hvacThermostat']);
        await reporting.thermostatTemperature(endpoint);
        await reporting.thermostatOccupiedHeatingSetpoint(endpoint);
        await reporting.thermostatUnoccupiedHeatingSetpoint(endpoint);
    },
};

```

It's working as expected but i have two questions:
1] All write operations works only few seconds after the device reported something. Should i set longer timeout, or is there a better way how to handle this?
2] There is already some enum and mapping for ctrlSeqeOfOper (cooling_only, cooling_with_reheat, etc...) what is the best way to handle this? Custom handler directly in that device class for both read / write? Any example?
3] I noticed that all stuff in repo are TS, do i have to use JS to define the device?

Thanks a lot.

**Comment by VladislavVesely** [comment]
Points 2] 3] resolved

But still i do not know what is the best way how to handle write which does time out before i'll receive something from the device.

**Comment by Koenkk** [comment]
1. Is it detected as and `end device`? If yes that's expected

**Reply by VladislavVesely** [reply]
Yes it is. But what is the solution then? Is it possible to wait for first event? Wake up the device? Prolong the write timeout?

**Reply by Koenkk** [reply]
Could you try adding the following to the definition? This will send the command once the device wakes-up

```js
        configure: async (device, coordinatorEndpoint) => {
            device.pendingRequestTimeout = 300000;
        },
```

**Reply by VladislavVesely** [reply]
> Could you try adding the following to the definition? This will send the command once the device wakes-up
> 
> ```js
>         configure: async (device, coordinatorEndpoint) => {
>             device.pendingRequestTimeout = 300000;
>         },
> ```

Does not seems to make a difference. On standalone installation it will failed after the timeout from write
await entity.write('hvacThermostat', {systemMode}, {timeout: 60000}); 

when i tried to move timeout to meta: {} or remove them all together and keep just that pendingRequest it does not have any effect at all. I tried also to set QOS levels but that does not have any effect either.


At the same time setting write timeout to some big value it will eventually fail as well probably on some different timeout level. Which is even more visible on HA installation where all configs are the same as the one i have standalone for testing (just different driver and zigbee stick). For example this one will fail after just 5 second totally ignoring everything.

```
Error 2024-05-29 09:16:59Publish 'set' 'system_mode' to 'TEST_SIEMENS_TERMOSTAT' failed: 'Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":300000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Delivery failed for {"profileId":260,"clusterId":513,"sourceEndpoint":1,"destinationEndpoint":1,"options":4416,"groupId":0,"sequence":33})'
Info 2024-05-29 09:17:02Received network/route error ROUTE_ERROR_INDIRECT_TRANSACTION_EXPIRY for "43478".
```

**Reply by Koenkk** [reply]
Could you try with: `await entity.write('hvacThermostat', {systemMode}, {sendPolicy: 'queue'});`

**Reply by VladislavVesely** [reply]
> Could you try with: `await entity.write('hvacThermostat', {systemMode}, {sendPolicy: 'queue'});`

No difference still getting  (Data request failed with error: 'MAC transaction expired' (240))

**Reply by Koenkk** [reply]
Could you provide the debug log when sending the command with this code?

See [this](https://www.zigbee2mqtt.io/guide/usage/debug.html) on how to enable debug logging.

**Reply by VladislavVesely** [reply]
> Could you provide the debug log when sending the command with this code?
> 
> See [this](https://www.zigbee2mqtt.io/guide/usage/debug.html) on how to enable debug logging.

That link is returning 404, so I used https://web.archive.org/web/20240525031346/https://www.zigbee2mqtt.io/guide/usage/debug.html but i guess there is no change in that at all. 

All the test are done on rpi5 running Debian Bookworm without desktop from service with  zstack ZBDongle-P (I do have also ember ZBDongle-E on x64 virtual pc running HA if needed for testing, because if does behaves differently).

Also keep in mind, that it is working if i hit the right window before the device sends something, so I intentionally trying to set something after the incoming message to be in the very worst position.

**1] Default write without any additional settings.**
`2024-06-01 18:19:30<-- [254,3,68,128,240,1,12,58]
Debug 2024-06-01 18:19:30--- parseNext [254,3,68,128,240,1,12,58]
Debug 2024-06-01 18:19:30--> parsed 3 - 2 - 4 - 128 - [240,1,12] - 58
Debug 2024-06-01 18:19:30AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":12}
Debug 2024-06-01 18:19:30--- parseNext []
Debug 2024-06-01 18:19:30Data confirm error (0x0080e12501677d35:50898,240,4)
Debug 2024-06-01 18:19:30ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))
Error 2024-06-01 18:19:30Publish 'set' 'system_mode' to '0x0080e12501677d35' failed: 'Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240))'
Debug 2024-06-01 18:19:30Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (Data request failed with error: 'MAC transaction expired' (240)) at ZStackAdapter.sendZclFrameToEndpointInternal (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts:412:23) at Queue.execute (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/utils/queue.ts:35:20) at Request.send (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/helpers/request.ts:79:20) at Endpoint.zclCommand (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:766:28) at Endpoint.write (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:358:9) at Object.convertSet (/opt/zigbee2mqtt/data/extension/externally-loaded.js:98:2) at Publish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/publish.ts:259:36) at EventEmitter.wrappedCallback (/opt/zigbee2mqtt/lib/eventBus.ts:174:17)`


**2] With send policy queue**

```
2024-06-01 18:22:29<-- [254,3,68,128,240,1,6,48]
Debug 2024-06-01 18:22:29--- parseNext [254,3,68,128,240,1,6,48]
Debug 2024-06-01 18:22:29--> parsed 3 - 2 - 4 - 128 - [240,1,6] - 48
Debug 2024-06-01 18:22:29AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":6}
Debug 2024-06-01 18:22:29--- parseNext []
Debug 2024-06-01 18:22:29Data confirm error (0x0080e12501677d35:50898,240,4)
Debug 2024-06-01 18:22:29ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))
Error 2024-06-01 18:22:29Publish 'set' 'system_mode' to '0x0080e12501677d35' failed: 'Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))'
Debug 2024-06-01 18:22:29Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":0}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240)) at ZStackAdapter.sendZclFrameToEndpointInternal (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts:412:23) at Queue.execute (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/utils/queue.ts:35:20) at Request.send (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/helpers/request.ts:79:20) at Endpoint.zclCommand (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:766:28) at Endpoint.write (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:358:9) at Object.convertSet (/opt/zigbee2mqtt/data/extension/externally-loaded.js:99:2) at Publish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/publish.ts:259:36) at EventEmitter.wrappedCallback (/opt/zigbee2mqtt/lib/eventBus.ts:174:17)
```

**3] With send policy queue and device pending request timeout**
```
2024-06-01 18:25:43<-- [254,3,68,128,240,1,8,62]
Debug 2024-06-01 18:25:43--- parseNext [254,3,68,128,240,1,8,62]
Debug 2024-06-01 18:25:43--> parsed 3 - 2 - 4 - 128 - [240,1,8] - 62
Debug 2024-06-01 18:25:43AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":8}
Debug 2024-06-01 18:25:43--- parseNext []
Debug 2024-06-01 18:25:43Data confirm error (0x0080e12501677d35:50898,240,4)
Debug 2024-06-01 18:25:43ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))
Error 2024-06-01 18:25:43Publish 'set' 'system_mode' to '0x0080e12501677d35' failed: 'Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))'
Debug 2024-06-01 18:25:43Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240)) at ZStackAdapter.sendZclFrameToEndpointInternal (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts:412:23) at Queue.execute (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/utils/queue.ts:35:20) at Request.send (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/helpers/request.ts:79:20) at Endpoint.zclCommand (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:766:28) at Endpoint.write (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:358:9) at Object.convertSet (/opt/zigbee2mqtt/data/extension/externally-loaded.js:99:2) at Publish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/publish.ts:259:36) at EventEmitter.wrappedCallback (/opt/zigbee2mqtt/lib/eventBus.ts:174:17)
```

**4] With send policy queue, device pending request timeout and write timeout**
`2024-06-01 18:27:45<-- [254,3,68,128,240,1,6,48]
Debug 2024-06-01 18:27:45--- parseNext [254,3,68,128,240,1,6,48]
Debug 2024-06-01 18:27:45--> parsed 3 - 2 - 4 - 128 - [240,1,6] - 48
Debug 2024-06-01 18:27:45AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":6}
Debug 2024-06-01 18:27:45--- parseNext []
Debug 2024-06-01 18:27:45Data confirm error (0x0080e12501677d35:50898,240,4)
Debug 2024-06-01 18:27:45ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":300000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))
Error 2024-06-01 18:27:45Publish 'set' 'system_mode' to '0x0080e12501677d35' failed: 'Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":300000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240))'
Debug 2024-06-01 18:27:45Error: ZCL command 0x0080e12501677d35/1 hvacThermostat.write({"systemMode":4}, {"timeout":300000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false,"sendPolicy":"queue"}) failed (Data request failed with error: 'MAC transaction expired' (240)) at ZStackAdapter.sendZclFrameToEndpointInternal (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/adapter/z-stack/adapter/zStackAdapter.ts:412:23) at Queue.execute (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/utils/queue.ts:35:20) at Request.send (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/helpers/request.ts:79:20) at Endpoint.zclCommand (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:766:28) at Endpoint.write (/opt/zigbee2mqtt/node_modules/zigbee-herdsman/src/controller/model/endpoint.ts:358:9) at Object.convertSet (/opt/zigbee2mqtt/data/extension/externally-loaded.js:99:2) at Publish.onMQTTMessage (/opt/zigbee2mqtt/lib/extension/publish.ts:259:36) at EventEmitter.wrappedCallback (/opt/zigbee2mqtt/lib/eventBus.ts:174:17)`

**Reply by Koenkk** [reply]
I expect that scenario 3 goes [here](https://github.com/Koenkk/zigbee-herdsman/blob/64ac386dca1d2be90fb1ad1a3b7003119890e651/src/controller/model/endpoint.ts#L299) which tries to send it first, and in case it fails adds it to the queue.

I think it's because the `pendingRequestTimeout` is not set, could you try with the following in your definition:

```js
        onEvent: async (type, data, device) => {
            device.pendingRequestTimeout = 300000;
        },
```

**Reply by VladislavVesely** [reply]
> I expect that scenario 3 goes [here](https://github.com/Koenkk/zigbee-herdsman/blob/64ac386dca1d2be90fb1ad1a3b7003119890e651/src/controller/model/endpoint.ts#L299) which tries to send it first, and in case it fails adds it to the queue.
> 
> I think it's because the `pendingRequestTimeout` is not set, could you try with the following in your definition:
> 
> ```js
>         onEvent: async (type, data, device) => {
>             device.pendingRequestTimeout = 300000;
>         },
> ```

It's there, but i just realise, that i need to reconfigure the device to make any difference. It does make a difference, however now i'm not getting any message from the device during this timeout period. So it just get stuck and that's it :/

`Debug 2024-06-02 16:21:49sendZclFrameToEndpointInternal 0x0080e12501677d35:15060/1 (0,1,1)
Debug 2024-06-02 16:21:49SREQ: --> AF - dataRequest - {"dstaddr":15060,"destendpoint":1,"srcendpoint":1,"clusterid":513,"transid":22,"options":0,"radius":30,"len":7,"data":{"type":"Buffer","data":[16,10,2,28,0,48,4]}}
Debug 2024-06-02 16:21:49--> frame [254,17,36,1,212,58,1,1,1,2,22,0,30,7,16,10,2,28,0,48,4,230]
Debug 2024-06-02 16:21:49<-- [254,1,100,1,0,100]
Debug 2024-06-02 16:21:49--- parseNext [254,1,100,1,0,100]
Debug 2024-06-02 16:21:49--> parsed 1 - 3 - 4 - 1 - [0] - 100
Debug 2024-06-02 16:21:49SRSP: <-- AF - dataRequest - {"status":0}
Debug 2024-06-02 16:21:49--- parseNext []
Debug 2024-06-02 16:21:57<-- [254,3,68,128,240,1,22,32]
Debug 2024-06-02 16:21:57--- parseNext [254,3,68,128,240,1,22,32]
Debug 2024-06-02 16:21:57--> parsed 3 - 2 - 4 - 128 - [240,1,22] - 32
Debug 2024-06-02 16:21:57AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":22}
Debug 2024-06-02 16:21:57--- parseNext []
Debug 2024-06-02 16:21:57Data confirm error (0x0080e12501677d35:15060,240,1)
Debug 2024-06-02 16:21:57SREQ: --> UTIL - assocGetWithAddress - {"extaddr":"0x0080e12501677d35","nwkaddr":15060}
Debug 2024-06-02 16:21:57--> frame [254,10,39,74,53,125,103,1,37,225,128,0,212,58,227]
Debug 2024-06-02 16:21:57<-- [254,36,103,74,212,58,91,0,1,4,0,0,0,1,170,0,5,0,0,0,0,0,255,0,60,0,0,227,59,0,0,1,255,0,0,0,0,0,0,0,242]
Debug 2024-06-02 16:21:57--- parseNext [254,36,103,74,212,58,91,0,1,4,0,0,0,1,170,0,5,0,0,0,0,0,255,0,60,0,0,227,59,0,0,1,255,0,0,0,0,0,0,0,242]
Debug 2024-06-02 16:21:57--> parsed 36 - 3 - 7 - 74 - [212,58,91,0,1,4,0,0,0,1,170,0,5,0,0,0,0,0,255,0,60,0,0,227,59,0,0,1,255,0,0,0,0,0,0,0] - 242
Debug 2024-06-02 16:21:57SRSP: <-- UTIL - assocGetWithAddress - {"nwkaddr":15060,"addridx":91,"noderelation":1}
Debug 2024-06-02 16:21:57--- parseNext []
Debug 2024-06-02 16:21:57assocRemove(0x0080e12501677d35)
Debug 2024-06-02 16:21:57SREQ: --> UTIL - assocRemove - {"ieeeadr":"0x0080e12501677d35"}
Debug 2024-06-02 16:21:57--> frame [254,8,39,99,53,125,103,1,37,225,128,0,38]
Debug 2024-06-02 16:21:57<-- [254,1,103,99,0,5]
Debug 2024-06-02 16:21:57--- parseNext [254,1,103,99,0,5]
Debug 2024-06-02 16:21:57--> parsed 1 - 3 - 7 - 99 - [0] - 5
Debug 2024-06-02 16:21:57SRSP: <-- UTIL - assocRemove - {"status":0}
Debug 2024-06-02 16:21:57--- parseNext []
Debug 2024-06-02 16:21:57sendZclFrameToEndpointInternal 0x0080e12501677d35:15060/1 (0,2,1)
Debug 2024-06-02 16:21:57SREQ: --> AF - dataRequest - {"dstaddr":15060,"destendpoint":1,"srcendpoint":1,"clusterid":513,"transid":23,"options":0,"radius":30,"len":7,"data":{"type":"Buffer","data":[16,10,2,28,0,48,4]}}
Debug 2024-06-02 16:21:57--> frame [254,17,36,1,212,58,1,1,1,2,23,0,30,7,16,10,2,28,0,48,4,231]
Debug 2024-06-02 16:21:57<-- [254,1,100,1,0,100]
Debug 2024-06-02 16:21:57--- parseNext [254,1,100,1,0,100]
Debug 2024-06-02 16:21:57--> parsed 1 - 3 - 4 - 1 - [0] - 100
Debug 2024-06-02 16:21:57SRSP: <-- AF - dataRequest - {"status":0}
Debug 2024-06-02 16:21:57--- parseNext []
Debug 2024-06-02 16:21:59<-- [254,3,68,128,205,1,23,28]
Debug 2024-06-02 16:21:59--- parseNext [254,3,68,128,205,1,23,28]
Debug 2024-06-02 16:21:59--> parsed 3 - 2 - 4 - 128 - [205,1,23] - 28
Debug 2024-06-02 16:21:59AREQ: <-- AF - dataConfirm - {"status":205,"endpoint":1,"transid":23}
Debug 2024-06-02 16:21:59--- parseNext []
Debug 2024-06-02 16:21:59Data confirm error (0x0080e12501677d35:15060,205,2)
Debug 2024-06-02 16:21:59assocAdd(0x0080e12501677d35)
Debug 2024-06-02 16:21:59SREQ: --> UTIL - assocAdd - {"ieeeadr":"0x0080e12501677d35","nwkaddr":15060,"noderelation":1}
Debug 2024-06-02 16:21:59--> frame [254,11,39,100,53,125,103,1,37,225,128,0,212,58,1,205]
Debug 2024-06-02 16:21:59<-- [254,1,103,100,0,2]
Debug 2024-06-02 16:21:59--- parseNext [254,1,103,100,0,2]
Debug 2024-06-02 16:21:59--> parsed 1 - 3 - 7 - 100 - [0] - 2
Debug 2024-06-02 16:21:59SRSP: <-- UTIL - assocAdd - {"status":0}
Debug 2024-06-02 16:21:59--- parseNext []
Debug 2024-06-02 16:21:59Discovering route to 15060
Debug 2024-06-02 16:21:59SREQ: --> ZDO - extRouteDisc - {"dstAddr":15060,"options":0,"radius":30}
Debug 2024-06-02 16:21:59--> frame [254,4,37,69,212,58,0,30,148]
Debug 2024-06-02 16:21:59<-- [254,1,101,69,0,33]
Debug 2024-06-02 16:21:59--- parseNext [254,1,101,69,0,33]
Debug 2024-06-02 16:21:59--> parsed 1 - 3 - 5 - 69 - [0] - 33
Debug 2024-06-02 16:21:59SRSP: <-- ZDO - extRouteDisc - {"status":0}
Debug 2024-06-02 16:21:59--- parseNext []
Debug 2024-06-02 16:22:02sendZclFrameToEndpointInternal 0x0080e12501677d35:15060/1 (0,3,1)
Debug 2024-06-02 16:22:02SREQ: --> AF - dataRequest - {"dstaddr":15060,"destendpoint":1,"srcendpoint":1,"clusterid":513,"transid":24,"options":0,"radius":30,"len":7,"data":{"type":"Buffer","data":[16,10,2,28,0,48,4]}}
Debug 2024-06-02 16:22:02--> frame [254,17,36,1,212,58,1,1,1,2,24,0,30,7,16,10,2,28,0,48,4,232]
Debug 2024-06-02 16:22:02<-- [254,1,100,1,0,100]
Debug 2024-06-02 16:22:02--- parseNext [254,1,100,1,0,100]
Debug 2024-06-02 16:22:02--> parsed 1 - 3 - 4 - 1 - [0] - 100
Debug 2024-06-02 16:22:02SRSP: <-- AF - dataRequest - {"status":0}
Debug 2024-06-02 16:22:02--- parseNext []
Debug 2024-06-02 16:22:09<-- [254,3,68,128,240,1,24,46]
Debug 2024-06-02 16:22:09--- parseNext [254,3,68,128,240,1,24,46]
Debug 2024-06-02 16:22:09--> parsed 3 - 2 - 4 - 128 - [240,1,24] - 46
Debug 2024-06-02 16:22:09AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":24}
Debug 2024-06-02 16:22:09--- parseNext []
Debug 2024-06-02 16:22:09Data confirm error (0x0080e12501677d35:15060,240,3)
Debug 2024-06-02 16:22:09Request network address of '0x0080e12501677d35'
Debug 2024-06-02 16:22:09SREQ: --> ZDO - nwkAddrReq - {"ieeeaddr":"0x0080e12501677d35","reqtype":0,"startindex":0}
Debug 2024-06-02 16:22:09--> frame [254,10,37,0,53,125,103,1,37,225,128,0,0,0,69]
Debug 2024-06-02 16:22:09<-- [254,1,101,0,0,100,254,21,69,255,0,0,0,0,128,0,0,0,0,0,53,125,103,1,37,225,128,0,212,58,0,171]
Debug 2024-06-02 16:22:09--- parseNext [254,1,101,0,0,100,254,21,69,255,0,0,0,0,128,0,0,0,0,0,53,125,103,1,37,225,128,0,212,58,0,171]
Debug 2024-06-02 16:22:09--> parsed 1 - 3 - 5 - 0 - [0] - 100
Debug 2024-06-02 16:22:09SRSP: <-- ZDO - nwkAddrReq - {"status":0}
Debug 2024-06-02 16:22:09--- parseNext [254,21,69,255,0,0,0,0,128,0,0,0,0,0,53,125,103,1,37,225,128,0,212,58,0,171]
Debug 2024-06-02 16:22:09--> parsed 21 - 2 - 5 - 255 - [0,0,0,0,128,0,0,0,0,0,53,125,103,1,37,225,128,0,212,58,0] - 171
Debug 2024-06-02 16:22:09AREQ: <-- ZDO - msgCbIncoming - {"srcaddr":0,"wasbroadcast":0,"clusterid":32768,"securityuse":0,"seqnum":0,"macdstaddr":0,"msgdata":{"type":"Buffer","data":[]}}
Debug 2024-06-02 16:22:09--- parseNext []
Debug 2024-06-02 16:22:19sendZclFrameToEndpointInternal 0x0080e12501677d35:15060/1 (0,4,1)
Debug 2024-06-02 16:22:19SREQ: --> AF - dataRequest - {"dstaddr":15060,"destendpoint":1,"srcendpoint":1,"clusterid":513,"transid":25,"options":0,"radius":30,"len":7,"data":{"type":"Buffer","data":[16,10,2,28,0,48,4]}}
Debug 2024-06-02 16:22:19--> frame [254,17,36,1,212,58,1,1,1,2,25,0,30,7,16,10,2,28,0,48,4,233]
Debug 2024-06-02 16:22:19<-- [254,1,100,1,0,100]
Debug 2024-06-02 16:22:19--- parseNext [254,1,100,1,0,100]
Debug 2024-06-02 16:22:19--> parsed 1 - 3 - 4 - 1 - [0] - 100
Debug 2024-06-02 16:22:19SRSP: <-- AF - dataRequest - {"status":0}
Debug 2024-06-02 16:22:19--- parseNext []
Debug 2024-06-02 16:22:27<-- [254,3,68,128,240,1,25,47]
Debug 2024-06-02 16:22:27--- parseNext [254,3,68,128,240,1,25,47]
Debug 2024-06-02 16:22:27--> parsed 3 - 2 - 4 - 128 - [240,1,25] - 47
Debug 2024-06-02 16:22:27AREQ: <-- AF - dataConfirm - {"status":240,"endpoint":1,"transid":25}
Debug 2024-06-02 16:22:27--- parseNext []
Debug 2024-06-02 16:22:27Data confirm error (0x0080e12501677d35:15060,240,4)
Debug 2024-06-02 16:22:27Request Queue (0x0080e12501677d35/1): queue request (transaction failed)
Debug 2024-06-02 16:22:27Request Queue (0x0080e12501677d35/1): Sending when active. Expires: 1717338400785
Debug 2024-06-02 16:22:50Saving state to file /opt/zigbee2mqtt/data/state.json`

**Reply by Koenkk** [reply]
The message should be send when the device wakes up, does this happen?

**Reply by VladislavVesely** [reply]
> The message should be send when the device wakes up, does this happen?

No. By default i can see a message received every one minute. Once i'll try to write this with the queue, i'll stop receiving everything or at least nothing in log.

**Reply by Koenkk** [reply]
Could you provide the debug log of:
- 5 minutes of doing nothing
- send request
- 5 minutes after this

**Reply by VladislavVesely** [reply]
Now it seems to be working as expected. Probably some part of configuration originally failed, there was no reporting configuration at all.

Thanks a lot for the support. One more thing. Is there way how to set common metadata for all write / read operations, so i can use just generic methods instead of copy pasting everything just to change the meta? 

Setting meta directly on root object does not seem to be doing anything.

Also do you think is worth of having this queue also for read operation?

EDIT: Or maybe not... i left the device and z2m for a while without making any changes. Tested the scenario again and see, that there is no queue in the log again. As if it does ignored the pending request timeout :/

**Reply by Koenkk** [reply]
We could add that to `utils.getOptions` which is used a lot already in `toZigbee`.

**Reply by VladislavVesely** [reply]
I meant it more generally, so we do not have to copy paste thermostat_system_mode just to add sendPolicy. I'm not sure how common it is for other implementations to be so close to original that only meta is different.

Btw what is the best process now to submit the device? In the doc you mentioned using JS, but whole repo is in TS. Is there a way how to test it from TS directly?

**Reply by Koenkk** [reply]
> I meant it more generally, so we do not have to copy paste thermostat_system_mode just to add sendPolicy. I'm not sure how common it is for other implementations to be so close to original that only meta is different.

Maybe I don't understand you, but this is exactly what `utils.getOptions` does

> Btw what is the best process now to submit the device? In the doc you mentioned using JS, but whole repo is in TS. Is there a way how to test it from TS directly?

You can submit a PR, for the definition part there is no diff between ts and js.
